megatron/core/optimizer/cpu_offloading/hybrid_optimizer.py

```diff
@@ -25,17 +30,30 @@ def register_grad_cpu_copy_hook(self):
         def grad_cpu_copy_hook_closure():
             def grad_cpu_copy_hook(optimizer, args, kwargs):
                 for gpu_param, cpu_copy in self.gpu_params_map_cpu_copy.items():
-                    cpu_copy.grad = gpu_param.grad.cpu()
-
+                    if cpu_copy not in self.cpu_copy_map_grad:
+                        self.cpu_copy_map_grad[cpu_copy] = torch.empty(
+                            gpu_param.grad.shape,
+                            dtype=gpu_param.grad.dtype,
+                            pin_memory=self.pin_cpu_grads
+                        )
+                self._data_stream.wait_stream(torch.cuda.default_stream())
+                with torch.cuda.stream(self._data_stream):
+                    for gpu_param, cpu_copy in self.gpu_params_map_cpu_copy.items():
+                        self.cpu_copy_map_grad[cpu_copy].data.copy_(gpu_param.grad, non_blocking=True)
+                        cpu_copy.grad = self.cpu_copy_map_grad[cpu_copy]
             return grad_cpu_copy_hook
 
         self.register_step_pre_hook(grad_cpu_copy_hook_closure())
 
     def register_param_copy_back_gpu_hook(self):
         def param_copy_back_gpu_hook_closure():
             def param_copy_back_gpu_hook(optimizer, args, kwargs):
-                for cpu_copy, gpu_param in self.cpu_copys_map_gpu_param.items():
-                    gpu_param.data.copy_(cpu_copy.data)
+                self._data_stream.wait_stream(torch.cuda.default_stream())
+                with torch.cuda.stream(self._data_stream):      
+                    for cpu_copy, gpu_param in self.cpu_copys_map_gpu_param.items():
+                        gpu_param.data.copy_(cpu_copy.data, non_blocking=True)
+                # NOTE: ensure all H2D/D2H Transfer and update operations finish
+                torch.cuda.synchronize()
```

Comment:
Can this be accomplished with a stream wait event to avoid the CPU overhead?

Reply:
In the most recent commit, we replace global synchronize with two record_event(), which blocks current stream until gpu_step() and cpu_copy_back() finish.